File: sdk/go/modules/stepper/stepper.go
// Package stepper provides a typed wrapper for the CONDUYT stepper-driver module.
package stepper

import (
	"context"
	"encoding/binary"

	conduyt "github.com/virgilvox/conduyt/sdk/go"
)

// Stepper provides typed control of a step/dir/enable stepper driver.
type Stepper struct {
	device   *conduyt.Device
	moduleID byte
}

// New creates a Stepper wrapper for the given device and module ID.
func New(device *conduyt.Device, moduleID byte) *Stepper {
	return &Stepper{device: device, moduleID: moduleID}
}

// Config sets the driver pins. Pass enPin=0xFF if the driver has no enable line.
func (s *Stepper) Config(ctx context.Context, stepPin, dirPin, enPin byte, stepsPerRev uint16) error {
	payload := make([]byte, 7)
	payload[0] = s.moduleID
	payload[1] = 0x01
	payload[2] = stepPin
	payload[3] = dirPin
	payload[4] = enPin
	binary.LittleEndian.PutUint16(payload[5:7], stepsPerRev)
	_, err := s.device.ModCmd(ctx, payload)
	return err
}

// Move runs the stepper a relative number of steps (negative reverses) at speedHz.
func (s *Stepper) Move(ctx context.Context, steps int32, speedHz uint16) error {
	payload := make([]byte, 8)
	payload[0] = s.moduleID
	payload[1] = 0x02
	binary.LittleEndian.PutUint32(payload[2:6], uint32(steps))
	binary.LittleEndian.PutUint16(payload[6:8], speedHz)
	_, err := s.device.ModCmd(ctx, payload)
	return err
}

// MoveTo runs the stepper to absolute position at speedHz.
func (s *Stepper) MoveTo(ctx context.Context, position int32, speedHz uint16) error {
	payload := make([]byte, 8)
	payload[0] = s.moduleID
	payload[1] = 0x03
	binary.LittleEndian.PutUint32(payload[2:6], uint32(position))
	binary.LittleEndian.PutUint16(payload[6:8], speedHz)
	_, err := s.device.ModCmd(ctx, payload)
	return err
}

// Stop halts the stepper immediately.
func (s *Stepper) Stop(ctx context.Context) error {
	_, err := s.device.ModCmd(ctx, []byte{s.moduleID, 0x04})
	return err
}

File: sdk/rust/src/modules/stepper.rs
//! CONDUYT Stepper Module — typed wrapper

use crate::device::{Device, DeviceError};
use crate::transports::Transport;

/// Stepper module wrapper. Drives a step/dir/enable stepper driver.
pub struct Stepper<'a, T: Transport> {
    device: &'a mut Device<T>,
    module_id: u8,
}

impl<'a, T: Transport> Stepper<'a, T> {
    pub fn new(device: &'a mut Device<T>, module_id: u8) -> Self {
        Self { device, module_id }
    }

    /// Configure pins. Pass `en_pin = 0xFF` if the driver has no enable line.
    pub fn config(&mut self, step_pin: u8, dir_pin: u8, en_pin: u8, steps_per_rev: u16)
        -> Result<(), DeviceError<T::Error>>
    {
        let p = [
            self.module_id, 0x01, step_pin, dir_pin, en_pin,
            (steps_per_rev & 0xFF) as u8, ((steps_per_rev >> 8) & 0xFF) as u8,
        ];
        self.device.mod_cmd(&p)?;
        Ok(())
    }

    /// Move a relative number of steps (negative reverses) at speedHz.
    pub fn r#move(&mut self, steps: i32, speed_hz: u16) -> Result<(), DeviceError<T::Error>> {
        let s = steps.to_le_bytes();
        let p = [
            self.module_id, 0x02,
            s[0], s[1], s[2], s[3],
            (speed_hz & 0xFF) as u8, ((speed_hz >> 8) & 0xFF) as u8,
        ];
        self.device.mod_cmd(&p)?;
        Ok(())
    }

    /// Move to absolute position at speedHz.
    pub fn move_to(&mut self, position: i32, speed_hz: u16) -> Result<(), DeviceError<T::Error>> {
        let s = position.to_le_bytes();
        let p = [
            self.module_id, 0x03,
            s[0], s[1], s[2], s[3],
            (speed_hz & 0xFF) as u8, ((speed_hz >> 8) & 0xFF) as u8,
        ];
        self.device.mod_cmd(&p)?;
        Ok(())
    }

    /// Halt the stepper immediately.
    pub fn stop(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x04])?;
        Ok(())
    }
}
